### PR TITLE
Add CMAESPolicy: (mu/mu_w, lambda)-CMA-ES over WeightedLinearPolicy weights

### LINKED ISSUE #19: Policy: CMA-ES (Covariance Matrix Adaptation Evolution Strategy)
## Summary

Replace the isotropic Gaussian mutation in `GeneticPolicy` with a dedicated **CMA-ES policy**. CMA-ES adapts a full covariance matrix over the weight space, allowing it to exploit correlations between features and converge far faster than per-weight Gaussian noise on ill-conditioned reward landscapes.

## Motivation

`GeneticPolicy` currently mutates weights with independent Gaussian noise (`mutation_scale`, `mutation_share`). This is equivalent to a diagonal-covariance ES with fixed variance. CMA-ES:

- Adapts step size (σ) via cumulative path length control
- Learns correlations between weight dimensions via rank-1 and rank-μ updates to the covariance matrix **C**
- Provides automatic step-size adaptation — no manual `mutation_scale` tuning
- Converges in O(n log n) function evaluations on smooth landscapes vs O(n²) for (1+λ)-ES

This is especially valuable for `WeightedLinearPolicy` whose weight vector has ~63 dimensions (21 obs × 3 heads).

## Proposed Implementation

Add `CMAESPolicy` to `policies.py`:

```python
class CMAESPolicy(BasePolicy):
    """
    CMA-ES over the flat weight vector of a WeightedLinearPolicy.
    Uses the (μ/μ_w, λ)-CMA-ES algorithm (Hansen 2016).
    """
```

### Algorithm sketch

```
Initialise:
  mean  = flat weights of champion (or random)
  σ     = initial_sigma (config, default 0.3)
  C     = identity (n×n)
  p_c, p_σ = 0  (evolution paths)
  λ     = population_size
  μ     = λ // 2
  w_i   = log(μ+0.5) - log(i)  for i=1..μ  (recombination weights, normalised)

Each generation:
  1. Sample λ offspring: z_k ~ N(0, C), x_k = mean + σ * z_k
  2. Evaluate each x_k as a WeightedLinearPolicy for one episode → reward r_k
  3. Sort by r_k descending; compute weighted mean update Δmean
  4. Update p_σ, p_c (cumulative step-size / covariance paths)
  5. Update σ via CSA (cumulative step-size adaptation)
  6. Update C via rank-1 + rank-μ update
  7. Eigen-decompose C periodically (every λ/(10·n) gens)
  8. Save champion (highest r_k ever seen) as WeightedLinearPolicy YAML
```

### Config additions to `training_params.yaml`

```yaml
policy_type: cmaes
policy_params:
  population_size: 20     # λ — number of offspring per generation
  initial_sigma: 0.3      # initial step size
  elite_fraction: 0.5     # μ/λ ratio
  covariance_update_freq: 1   # generations between eigen-decompositions
```

## Files to change

| File | Change |
|------|--------|
| `policies.py` | Add `CMAESPolicy` class (~150 lines) |
| `config/training_params.yaml` | Document new `policy_type: cmaes` |
| `config/policy_weights.yaml` | No change (champion saved as `WeightedLinearPolicy` format) |
| `main.py` `_make_policy()` | Register `"cmaes"` case |
| `tests/test_policies.py` | Unit test: 1-D quadratic convergence in <50 generations |

## Acceptance Criteria

- [ ] `CMAESPolicy` inherits `BasePolicy`, implements `__call__`, `to_cfg`, `save`
- [ ] Champion saved in same YAML format as `WeightedLinearPolicy` (compatible with analytics)
- [ ] `policy_type: cmaes` works end-to-end in `main.py`
- [ ] Unit test verifies convergence on a simple quadratic reward function
- [ ] Grid search (`grid_search.py`) works when `policy_type: cmaes`

## References

- Hansen, N. (2016). *The CMA Evolution Strategy: A Tutorial.* arXiv:1604.00772
- Pure-numpy reference: https://github.com/CyberAgentAILab/cmaes (MIT)

### PR DESCRIPTION
**Pull request by espenhk**
Closes #19.

CMA-ES adapts a full covariance matrix over the flat [steer|accel|brake] weight vector (~63-D), providing automatic step-size adaptation and correlation learning without manual mutation tuning.

## Changes

- **`policies.py`**: `CMAESPolicy` class implementing `(μ/μ_w, λ)-CMA-ES` (Hansen 2016):
  - `sample_population()` / `update_distribution()` interface matching `GeneticPolicy`'s training loop contract
  - Cumulative step-size adaptation (CSA) for σ, rank-1 + rank-μ covariance update
  - Seeded `self._rng` for reproducible sampling (accepts optional `seed` parameter)
  - Public properties `population_size` and `sigma`
  - Guard in `update_distribution()` raises `RuntimeError` if called before a matching `sample_population()`
  - `save()` writes champion in `WeightedLinearPolicy` YAML format for analytics compatibility

- **`main.py`**: `_greedy_loop_cmaes()`, `_make_policy()` `'cmaes'` branch, dispatch in `train_rl()`

- **`config/training_params.yaml`**: documents `cmaes` `policy_params` block

- **`tests/test_cmaes_policy.py`**: 24 unit tests covering init, sampling, distribution updates, serialisation, guard behaviour, and deterministic quadratic convergence (fixed seed)

- **`README.md`** / **`CLAUDE.md`**: CMAESPolicy added to policy tables with algorithm description, hyperparameter reference, and usage notes

https://claude.ai/code/session_01Mj7MA8eay11RaGgMYd5XEJ

### CODE REVIEW

File: policies.py [hunk 1]
```diff
@@ -807,3 +808,242 @@ def save(self, path: str) -> None:
...
+        # Rank offspring descending by reward
+        order = np.argsort(rewards)[::-1]
+
+        # Update champion
+        improved = False
+        best_r   = rewards[order[0]]
+        if best_r > self._champion_reward:
+            self._champion_reward = best_r
+            self._champion        = self._flat_to_policy(self._pop_xs[order[0]])
+            improved              = True
```
[the remark this answers is not in the document]
Comment (espenhk) [reply policies.py:1022]: @copilot apply changes based on [this feedback](https://github.com/espenhk/tmnf-ai/pull/25#discussion_r3067699336)

File: main.py [hunk 2]
```diff
@@ -702,6 +720,89 @@ def _greedy_loop_genetic(
...
+
+    try:
+        for gen in range(1, n_generations + 1):
+            env._max_episode_time_s = _scaled_episode_time(gen, n_generations, full_episode_time_s)
+            population = policy.sample_population()
+            pop_size   = len(population)
+            logger.info(
+                "--- Generation %d/%d --- evaluating %d individuals (σ=%.4f, episode_time=%.1fs)",
+                gen, n_generations, pop_size, policy._sigma, env._max_episode_time_s,
+            )
```
[the remark this answers is not in the document]
Comment (espenhk) [reply main.py:761]: @copilot apply changes based on [this feedback](https://github.com/espenhk/tmnf-ai/pull/25#discussion_r3067699343)

File: tests/test_cmaes_policy.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+class TestCMAESConvergence(unittest.TestCase):
+
+    def test_converges_toward_quadratic_maximum(self):
+        """CMA-ES mean should move toward the maximizer of a quadratic in <=50 generations."""
+        from obs_spec import BASE_OBS_DIM
+
+        policy  = CMAESPolicy(population_size=20, initial_sigma=1.0, n_lidar_rays=0)
+        policy.initialize_random()
+
```
[the remark this answers is not in the document]
Comment (espenhk) [reply tests/test_cmaes_policy.py:212]: @copilot apply changes based on [this feedback](https://github.com/espenhk/tmnf-ai/pull/25#discussion_r3067699358)
